wts/blueprints/external_oidc.py

```diff
@@ -0,0 +1,144 @@
+    """
+
+    unexpired_only = flask.request.args.get("unexpired", "false").lower() == "true"
+
+    global external_oidc_cache
+    if not external_oidc_cache:
+        data = {
+            "providers": [
+                {
+                    # name to display on the login button
+                    "name": idp_conf["name"],
+                    # unique ID of the configured identity provider
+                    "idp": idp,
+                    # hostname URL - gen3fuse uses it to get the manifests
+                    "base_url": oidc_conf["base_url"],
+                    # authorization URL to use for logging in
+                    "urls": [
+                        {
+                            "name": idp_conf["name"],
+                            "url": generate_authorization_url(idp),
+                        }
+                    ],
+                }
+                for oidc_conf in get_config_var("EXTERNAL_OIDC", [])
+                for idp, idp_conf in oidc_conf.get("login_options", {}).items()
+            ]
+        }
+        external_oidc_cache = data
+
+    # get the username of the current logged in user.
+    # `current_user` validates the token and relies on `OIDC_ISSUER`
+    # to know the issuer
+    client, _ = get_oauth_client(idp="default")
+    flask.current_app.config["OIDC_ISSUER"] = client.api_base_url.strip("/")
+    username = None
+    try:
+        user = current_user
+        username = user.username
+    except Exception:
+        flask.current_app.logger.info(
+            "no logged in user: will return is_connected=False for all IDPs"
+        )
+
+    # get all expirations at once (1 DB query)
+    idp_to_token_exp = get_refresh_token_expirations(
+        username, [p["idp"] for p in external_oidc_cache["providers"]]
+    )
+
+    result = {"providers": []}
+    for p in external_oidc_cache["providers"]:
+        # expiration of the current user's refresh token
+        exp = idp_to_token_exp[p["idp"]]
+        if exp or not unexpired_only:
+            p["refresh_token_expiration"] = exp
+            result["providers"].append(p)
+
+    return flask.jsonify(result), 200
+
+
+def generate_authorization_url(idp):
+    """
+    Args:
+        idp (string)
+
+    Returns:
+        str: authorization URL to go through the OIDC flow and get a
+            refresh token for this IDP
+    """
+    wts_base_url = get_config_var("WTS_BASE_URL")
+    authorization_url = wts_base_url + "oauth2/authorization_url?idp=" + idp
+    return authorization_url
+
+
+def seconds_to_human_time(seconds):
+    if seconds < 0:
+        return None
+    m, s = divmod(seconds, 60)
+    h, m = divmod(m, 60)
+    d, h = divmod(h, 24)
+    if d:
+        return "{} days".format(d)
+    if h:
+        return "{} hours".format(h)
+    if m:
+        return "{} minutes".format(m)
+    return "{} seconds".format(s)
+
+
+def get_refresh_token_expirations(username, idps):
```

Comment:
I think we should also either have unit tests for this function, or be sure to test the behavior of the expiration tokens in unit tests for `get_external_oidc`. Specifically I think the behavior u should test is the 'return None if the most recent token out of this set of tokens is expired', because the code that decides that is complex.

Reply:
🤦‍♂Just realized -- do we not have unit tests at all in wts? I still think we really should add unit tests when we can -- especially since this code is related to auth.

Reply:
yeah there aren't any tests... i've been working on writing some in between other tasks, but i thought i'd submit this PR for review first since it's somewhat urgent and it's already big enough as it is